port1.0: Implement use_xcode flag

satraul opened the pull request:
To my understanding, MacPorts runs fine using only the Command Line Tools and this has been a problem for quite some time. In this PR, as part of the GSOC program, I have implemented several changes as the first step in providing a smooth experience of using MacPorts without Xcode:

1. In addition to the `port 1.0` PortGroup, maintainers can declare Xcode dependency using a new Portfile option `use_xcode`
2. Unless ports declare Xcode dependency:
    - Default ‘DEVELOPER_DIR’ to point Command Line Tools. Thus, shim Xcode binaries such as /usr/bin/clang and /usr/bin/gcc will point to Command Line Tools instead of Xcode. This will ensure all Xcode-independent ports have reproducible builds
    - Installing/building with trace mode will hide the Xcode installation, so MacPorts can fail builds for ports that actually need Xcode but are not declared as such. This will help maintainers identify Xcode dependency.
3. Error out in the `fetch` phase when users try to install Xcode-dependent ports without installing Xcode (if `$xcodeversion eq "none"`)

Relevant:
- https://trac.macports.org/ticket/35045
- https://trac.macports.org/ticket/58016

Thank you.

mojca commented on src/port1.0/portconfigure.tcl:
```diff
@@ -138,7 +138,7 @@ default compiler.fortran_fallback    {[portconfigure::get_fortran_fallback]}
 commands configure autoreconf automake autoconf xmkmf
 # defaults
-default configure.env       ""
+default configure.env       "DEVELOPER_DIR = /Library/Developer/CommandLineTools"
```
What's the compatibility of this with older OS versions? Do we set this on Linux as well? @kencu

Not that I expect from @satraul to have access to older versions, or test on every single version, but it would be helpful if someone could confirm that things don't break on older boxes.

neverpanic commented on src/port1.0/portbuild.tcl:
```diff
@@ -86,8 +88,12 @@ options build.type.add_deps
         return [findBinary make $portutil::autoconf::make_path]
+    }]
+    if {[option use_xcode] eq "yes" && $xcodebuildcmd eq "none"} {
+        return -code error "This port requires Xcode, which was not found on your system!"
```
That's probably not a good place to do this. This will cause an error when users that do not have Xcode installed try to parse the Portfile, and the port will hence completely disappear from their system because `portindex` will fail to find the details of this port.

Instead, you should do this check in one of the phases, e.g. in a pre-fetch block.

neverpanic commented on src/port1.0/portbuild.tcl:
```diff
@@ -91,8 +91,9 @@ proc portbuild::build_getmaketype {args} {
-    if {[option use_xcode] eq "yes" && $xcodebuildcmd eq "none"} {
+    }
+    # TODO: Implement reliable way of checking for Xcode
+    if {([option build.type] eq "xcode" || [option use_xcode] eq "yes") && $xcodebuildcmd eq "none"} {
```
Didn't you replace this with changing the default value of `use_xcode` if `build.type` is `xcode`? If so, shouldn't this commit be dropped from the history of the PR?

neverpanic commented on src/port1.0/porttrace.tcl:
```diff
@@ -136,6 +136,7 @@ namespace eval porttrace {
 
         variable fifo
         variable fifo_mktemp_template
+        set PLACEHOLDER_DENY_XCODE no
```
I'd prefer if this WIP change vanished from the history and you interactively rebased the final solution into this place.

neverpanic commented on doc/portfile.7:
```diff
@@ -1589,8 +1589,9 @@ Target passed to
-Specifies dependency on Xcode compilers. Restricts Xcode compilers on trace mode
-when set to no.
+Specifies dependency on Xcode compilers. Restricts Xcode compilers on trace
+mode when set to no. This option is overwritten to yes when build.type is
```
As mentioned on a different commit, this should probably indicate that only the default changes and developers can still set this value.

neverpanic commented on doc/portfile.7:
```diff
@@ -1588,6 +1588,18 @@ Target passed to
 .Sy Example:
 .Dl build.target all-src
+.It Ic use_xcode
+Specifies dependency on Xcode compilers. Restricts Xcode compilers on trace mode
```
I'd squash this with the following commit.

neverpanic commented on src/port1.0/porttrace.tcl:
```diff
@@ -35,6 +35,18 @@ package provide porttrace 1.0
 package require Pextlib 1.0
 package require portutil 1.0
 
+options use_xcode
```
As mentioned on a different commit, this should probably not be duplicated.

neverpanic commented on src/port1.0/portbuild.tcl:
```diff
@@ -49,6 +49,7 @@ options build.asroot \
         use_xcode
 commands build
 # defaults
+default build.env "DEVELOPER_DIR = /Library/Developer/CommandLineTools"
```
As mentioned on a different commit, there should be no space here.

satraul committed:
port1.0: Implement use_xcode option

satraul committed:
doc: Document use_xcode

satraul commented on src/port1.0/portutil.tcl:
```diff
@@ -508,6 +508,16 @@ proc default_check {optionName index op} {
 
+# env_default 
+# Unless port is Xcode dependent, returns env which sets developer_dir to Command Line Tools
+proc env_default {} {
```
@mojca I'll have to investigate that, but we can put some conditionals here if the path's different across versions.

satraul commented on src/port1.0/portfetch.tcl:
```diff
@@ -607,11 +607,20 @@ proc portfetch::fetch_addfilestomap {filemapname} {
     variable fetch_urls
 
     portfetch::checkfiles fetch_urls
+    portfetch::checkxcode
```
@neverpanic By "pre-fetch" do you mean like this? I don't fully understand the portindex problem yet.

neverpanic replied:
Yes, that'll work.

As for the portindex problem, try the following:
1. Create a local Portfile repository as described in https://guide.macports.org/#development.local-repositories
2. Write a port that contains `error "testing"` outside of a phase block
3. Run `portindex` as described in the guide. It will fail to index the port, and the generated Portindex will not contain any information about that port.

Now, change that port such that the `error "testing"` statement is in a `pre-fetch { ... }` block. If you run portindex again, it should successfully parse the port and add an entry to the Portindex. You still won't be able to install the port (because it's fetch phase will fail), but you will be able to run `port info` on it.

satraul commented:
So I created tested the new changes on a modified [emptyport](https://github.com/satraul/macports-base/blob/master/src/port1.0/tests/Portfile) with
```tcl
use_xcode           no
...
configure {
    ui_msg "CONFIGURE"
    ui_msg "========="
    ui_msg {OUTPUT OF ${configure.env}}
    ui_msg ${configure.env} 
    ui_msg
    ui_msg {OUTPUT OF [exec -ignorestderr clang -v]}
    ui_msg [exec -ignorestderr clang -v]
}
```
Then I ran `port configure` which output:
```
Portfile changed since last build; discarding previous state.
--->  Fetching distfiles for emptyport
--->  Verifying checksums for emptyport
--->  Extracting emptyport
--->  Configuring emptyport
CONFIGURE
=========
OUTPUT OF ${configure.env}
DEVELOPER_DIR=/Library/Developer/CommandLineTools

OUTPUT OF [exec -ignorestderr clang -v]
Apple LLVM version 10.0.0 (clang-1000.11.45.5)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Am I doing it wrong? I suspect that I shouldn't be calling clang with `exec`. The output should be:
```
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

satraul commented:
So I created tested the new environments on a modified [emptyport](https://github.com/satraul/macports-base/blob/master/src/port1.0/tests/Portfile) with
```tcl
use_xcode           no
...
configure {
    ui_msg "CONFIGURE"
    ui_msg "========="
    ui_msg {OUTPUT OF ${configure.env}}
    ui_msg ${configure.env} 
    ui_msg
    ui_msg {OUTPUT OF [exec -ignorestderr clang -v]}
    ui_msg [exec -ignorestderr clang -v]
}
```
Then I ran `port configure` which output:
```
Portfile changed since last build; discarding previous state.
--->  Fetching distfiles for emptyport
--->  Verifying checksums for emptyport
--->  Extracting emptyport
--->  Configuring emptyport
CONFIGURE
=========
OUTPUT OF ${configure.env}
DEVELOPER_DIR=/Library/Developer/CommandLineTools

OUTPUT OF [exec -ignorestderr clang -v]
Apple LLVM version 10.0.0 (clang-1000.11.45.5)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Am I doing it wrong? I suspect that I shouldn't be calling clang with `exec`. The output should be:
```
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

neverpanic commented on src/port1.0/portutil.tcl:
```diff
@@ -508,6 +508,16 @@ proc default_check {optionName index op} {
+# Unless port is Xcode dependent, returns env which sets developer_dir to Command Line Tools
+proc env_default {} {
+    global use_xcode
+    if {[info exists use_xcode] && $use_xcode eq "yes"} {
```
I don't think you need the `[info exists use_xcode]` here, since it's already defined as an option.

satraul replied:
Removed

neverpanic commented:
The reason why your code doesn't do what you expect it to do is that `configure.env` is only taken into account when running things via `configure.cmd`. So if you set `configure.cmd clang -v` and do not completely override the `configure` phase, you should see it working. You can get the value of `configure.env` in this setup by running `port -d`, since that always prints the environment.

If you want to follow along with where this happens, check for `command_exec configure` in `portconfigure.tcl`, which is implemented in `portutil.tcl`, which puts the values from `configure.env` into the environment before calling `system`.

satraul commented:
Right, thank you. I've previously read the `command_exec` proc but I couldn't figure out how to invoke it correctly. It's working as intended after changing it to `configure.cmd clang -v` on `use_xcode no`:

```
:debug:configure DEVELOPER_DIR='/Library/Developer/CommandLineTools'
...
:debug:configure system:  cd "/opt/macports-test/var/macports/build/_Users_firelink_git_macports-test/emptyport/work/emptyport-0.2" && clang -v --prefix=/opt/macports-test 
:info:configure Apple LLVM version 10.0.0 (clang-1000.10.44.4)
:info:configure Target: x86_64-apple-darwin17.7.0
:info:configure Thread model: posix
:info:configure InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

On `use_xcode yes`:

```
:info:configure Executing:  cd "/opt/macports-test/var/macports/build/_Users_firelink_git_macports-test/emptyport/work/emptyport-0.2" && clang -v --prefix=/opt/macports-test 
:debug:configure system:  cd "/opt/macports-test/var/macports/build/_Users_firelink_git_macports-test/emptyport/work/emptyport-0.2" && clang -v --prefix=/opt/macports-test 
:info:configure Apple LLVM version 10.0.0 (clang-1000.11.45.5)
:info:configure Target: x86_64-apple-darwin17.7.0
:info:configure Thread model: posix
:info:configure InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

satraul commented on src/port1.0/porttrace.tcl:
```diff
@@ -213,18 +213,29 @@ namespace eval porttrace {
-        allow trace_sandbox [file normalize ~${macportsuser}/Library/Preferences/com.apple.dt.Xcode.plist]
-        allow trace_sandbox "$env(HOME)/Library/Preferences/com.apple.dt.Xcode.plist"
+        set xcode_paths {}
+        lappend xcode_paths "/var/db/xcode_select_link"
```
Not sure which ones should be hidden or not

neverpanic replied:
Unfortunately I don't know either. You can do the following to find out:

Modify your `configure.cmd clang -v` Portfile to explicitly set `configure.env DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer` (and leave `use_xcode` at the default of `no`). If you run this Portfile in trace mode, it should fail. If it doesn't, we're not filtering the right things yet, or the filtering doesn't work.

neverpanic replied:
Actually, we should probably also include the entire Xcode installation location in the sandbox. Nothing should be used from `/Applications/Xcode.app/Contents/Developer` with `use_xcode no`.

satraul replied:
On Portfile:
```tcl
configure.env       DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer
configure.cmd       clang -v
```
`port configure -t` correctly errors out with:
```
:info:configure Executing:  cd "/opt/macports-test/var/macports/build/_Users_firelink_git_macports-test/emptyport/work/emptyport-0.2" && clang -v --prefix=/opt/macports-test 
:debug:configure system:  cd "/opt/macports-test/var/macports/build/_Users_firelink_git_macports-test/emptyport/work/emptyport-0.2" && clang -v --prefix=/opt/macports-test 
:info:configure xcrun: error: invalid DEVELOPER_DIR path (/Applications/Xcode.app/Contents/Developer), missing xcrun at: /Applications/Xcode.app/Contents/Developer/usr/bin/xcrun
:info:configure Command failed:  cd "/opt/macports-test/var/macports/build/_Users_firelink_git_macports-test/emptyport/work/emptyport-0.2" && clang -v --prefix=/opt/macports-test 
:info:configure Exit code: 1
:info:configure The following file would have been hidden from the build system by trace mode if it existed:
:info:configure   /Applications/Xcode.app/Contents/Developer/mach_kernel
:error:configure Failed to configure emptyport: configure failure: command execution failed
:debug:configure Error code: NONE
:debug:configure Backtrace: configure failure: command execution failed
:debug:configure     while executing
:debug:configure "$procedure $targetname"
:error:configure See /opt/macports-test/var/macports/logs/_Users_firelink_git_macports-test/emptyport/main.log for details.
```
So I think `/Applications/Xcode.app/Contents/Developer` is already hidden in the sandbox courtesy of this code:
```tcl
set ddsplit [file split [file normalize [file join ${developer_dir} ".." ".."]]]
if {[llength $ddsplit] > 2 && [lindex $ddsplit end-1] eq "Contents" && [lindex $ddsplit end] eq "Developer"} {
    set ddsplit [lrange $ddsplit 0 end-2]
}
lappend xcode_paths [file join {*}$ddsplit]
```
I also tested that trace mode doesn't hide CommandLineTools. On the same Portfile with only `configure.cmd       clang -v`, `port configure -t` runs successfully so it should be OK.

satraul committed:
Check Xcode existence by $xcodeversion

Following https://github.com/macports/macports-base/blob/1d6588d178b251c86abc37cbea5ab580e2b85fd1/src/port1.0/portutil.tcl#L3296

satraul commented:
Additionally tested modified HexFiend (`xcode 1.0 PortGroup`) with `use_xcode no` for `port install -t` which errors out correctly:

```
:info:build xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
:info:build Command failed: cd "/opt/macports-test/var/macports/build/_Users_firelink_git_macports-ports_editors_HexFiend/HexFiend/work/HexFiend-2.9.0" && DEVELOPER_DIR=/Library/Developer/CommandLineTools /usr/bin/xcodebuild  -target "HexFiend_Framework" -configuration Release build INSTALL_PATH=/opt/macports-test/Applications  OBJROOT="/opt/macports-test/var/macports/build/_Users_firelink_git_macports-ports_editors_HexFiend/HexFiend/work/HexFiend-2.9.0/build/" SYMROOT="/opt/macports-test/var/macports/build/_Users_firelink_git_macports-ports_editors_HexFiend/HexFiend/work/HexFiend-2.9.0/build/" MACOSX_DEPLOYMENT_TARGET=10.8 ARCHS=x86_64 SDKROOT="" GCC_VERSION=com.apple.compilers.llvm.clang.1_0 CLANG_CXX_LIBRARY="libc++" 
:info:build Exit code: 1
:error:build Failed to build HexFiend: command execution failed
:debug:build Error code: NONE
:debug:build Backtrace: command execution failed
:debug:build     while executing
:debug:build "$procedure $targetname"
:error:build See /opt/macports-test/var/macports/logs/_Users_firelink_git_macports-ports_editors_HexFiend/HexFiend/main.log for details.
```

satraul commented on src/port1.0/porttrace.tcl:
```diff
@@ -213,18 +213,29 @@ namespace eval porttrace {
-        allow trace_sandbox "$env(HOME)/Library/Preferences/com.apple.dt.Xcode.plist"
+        set xcode_paths {}
+        lappend xcode_paths "/var/db/xcode_select_link"
+        lappend xcode_paths "/var/db/mds"
```
The `/var/db/xcode_select_link` symlink is what `libxcselect` reads if DEVELOPER_DIR is not set. I wrote about this in my GSOC proposal. This could lead to Xcode.app so restricting it is needed.
 https://macops.ca/developer-binaries-on-os-x-xcode-select-and-xcrun/

After some searching.. I think `/var/db/mds` is for Spotlight.
Not sure if we need to deny access to it. But since it's here in the first place (I assume to search for Xcode), I'll leave it as such.
https://apple.stackexchange.com/questions/238776/what-is-the-purpose-of-mds-folder-on-a-mac

satraul commented:
Moved PR to https://github.com/macports/macports-base/pull/130
